### PR TITLE
[amp refactors 2/3] Refactor 'parent required but nonexistent' condition into rule

### PR DESCRIPTION
**Pull request by clairelin135**
Refactors the 'parent required but nonexistent' condition into a skip rule.

Adds a custom rule evaluation data type that contains asset keys with required but nonexistent partitions. We could also provide the specific partitions that are required but nonexistent, but we don't provide information at this granularity for the remainder of the rule evaluation data types.

Adds tests for the new skip rule to explicitly test behavior.

### CODE REVIEW

**Comment by clairelin135** [comment]
Current dependencies on/for this PR:
* master
  * **PR #17142** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/17142" target="_blank">Graphite</a> 
    * **PR #17144** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/17144" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/17144?utm_source=stack-comment).

File: js_modules/dagster-ui/packages/ui-core/src/assets/AutoMaterializePolicyPage/RuleEvaluationOutcomes.tsx [hunk 1]
```diff
@@ -147,6 +147,8 @@ const RightElementForEvaluations = ({
       );
     case 'WaitingOnKeysRuleEvaluationData':
       return <WaitingOnAssetKeysLink assetKeys={first.waitingOnAssetKeys || []} />;
+    case 'RequiredButNonexistentParentsRuleEvaluationData':
```
Comment (clairelin135) [inline js_modules/dagster-ui/packages/ui-core/src/assets/AutoMaterializePolicyPage/RuleEvaluationOutcomes.tsx:150]: For now display nothing on the right hand side. This looks like:
image

**Review by OwenKephart — DISMISSED** [review]
Just at a high level, I think we should avoid adding anything to the graphql layer if at all possible. In this instance, I think `WaitingOnKeysRuleEvaluationData` would apply pretty well to the use case we have here (we're waiting on certain keys to have partitions available).

This way, we get nice UI treatement out of the box

**Comment by clairelin135** [comment]
Agree that it's ideal to not have to support additional graphQL types. 

In most cases that this rule fires, I don't think that we're waiting for partitions though -- those partitions generally will never exist. Regardless, I'm fine with continuing to use `WaitingOnKeysRuleEvaluationData` and refactoring it to have a more open-ended name at a later time, especially since we don't plan on returning specific partitions

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py [hunk 2]
```diff
@@ -624,6 +634,47 @@ def evaluate_for_asset(self, context: RuleEvaluationContext) -> RuleEvaluationRe
...
+class SkipOnRequiredButNonexistentParentsRule(
+    AutoMaterializeRule, NamedTuple("_SkipOnRequiredButNonexistentParentsRule", [])
+):
+    @property
+    def decision_type(self) -> AutoMaterializeDecisionType:
+        return AutoMaterializeDecisionType.SKIP
+
+    @property
+    def description(self) -> str:
+        return "parents are required but do not exist"
```
Comment (OwenKephart) [inline python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py:647]: maybe something that references partitions here would be a bit more informative. I was thinking something along the lines of

"required parent partitions do not exist"

The only concern there is just that people might confuse "not exist" with "have not been materialized" -- curious if you have thoughts on a better term / phrase there

**Review by OwenKephart — COMMENTED** [review]
Overall, I'm on board!

Just double checking -- what happens in cases where this rule is not on a policy, but we do run into this situation?

My expectation is that it successfully creates a run, but executing that run would result in a failure (rather than the daemon itself erroring). Is that right?

**Comment by clairelin135** [comment]
> My expectation is that it successfully creates a run, but executing that run would result in a failure (rather than the daemon itself erroring)

Yup, this is the behavior that will happen

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py (hunk 2 above)
Comment (clairelin135) [reply python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py]: I think this description looks clear to me.

I don't think people would be confused about this language, but we can also add a more verbose tooltip (i.e. `required upstream partitions are not defined in upstream asset X`)

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py [hunk 3]
```diff
@@ -286,6 +286,16 @@ def skip_on_not_all_parents_updated(
         """
         return SkipOnNotAllParentsUpdatedRule(require_update_for_all_parent_partitions)
 
+    @public
+    @staticmethod
+    def skip_on_required_but_nonexistent_parents() -> "SkipOnRequiredButNonexistentParentsRule":
+        """Skip an asset partition if it depends on parent partitions that do not exist.
```
Comment (OwenKephart) [inline python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py:292]: Final nit on cleaning up this docstring, maybe something along the lines of "For example, imagine a downstream asset is time-partitioned, starting in 2022, but has a time-partitioned parent which starts in 2023. This rule will skip attempting to materialize downstream partitions from before the start date of the parent"

**Review by OwenKephart — APPROVED** [review]
note nit on docstring, otherwise great 👍

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py (hunk 3 above)
Comment (clairelin135) [reply python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py:292]: updated!

**Pull request merged**